### PR TITLE
Add floating-point comparison example passes

### PR DESCRIPTION
**Pull request by jvstech**
This adds an analysis pass and a transformation pass that act on instructions performing direct floating-point value comparisons.

### CODE REVIEW

**Review by banach-space — COMMENTED** [review]
Hi @jvstech, thank you for submitting this! 

It looks very neat, is thoroughly documented and tested. And it's a very exciting and relatable example - I personally like it a lot :) This is great work! It's the first example in **llvm-tutor** that covers:

* floating point numbers
* printing passes for the new PM

It's also a neat example of passes depending on other passes (majority of examples so far don't require other passes to run). I've left a few comments, but in most cases it's just fine-tuning and please consider them as [nit]s. 

It would be great to do a bit more testing. For example, what about other predicates? Perhaps you could add this file:

```llvm
; RUN: opt -load %shlibdir/libFindFCmpEq%shlibext  -load %shlibdir/libConvertFCmpEq%shlibext -convert-fcmp-eq  -S %s \
; RUN:  | FileCheck %s
; RUN: opt -load-pass-plugin=%shlibdir/libFindFCmpEq%shlibext  -load-pass-plugin=%shlibdir/libConvertFCmpEq%shlibext --passes=convert-fcmp-eq  -S %s \
; RUN:  | FileCheck %s

define i32 @fp_cmp(double %a, double %b) {
; CHECK-LABEL: @fp_cmp
; CHECK-DAG: %1 = fsub double %a, %b
; CHECK-NEXT: %2 = bitcast double %1 to i64
; CHECK-NEXT: %3 = and i64 %2, 9223372036854775807
; CHECK-NEXT: %4 = bitcast i64 %3 to double
; CHECK-NEXT: %cmp = fcmp olt double %4, 0x3CB0000000000000
; CHECK-NEXT: %conv = zext i1 %cmp to i32
; CHECK-NOT: fcmp oeq
; CHECK-DAG: ret i32 %conv

  %cmp = fcmp oeq double %a, %b
  %conv = zext i1 %cmp to i32
  ret i32 %conv
}

define i32 @fp_cmp1(double %a, double %b) {
; CHECK-LABEL: @fp_cmp1
; CHECK-DAG: %1 = fsub double %a, %b
; CHECK-NEXT: %2 = bitcast double %1 to i64
; CHECK-NEXT: %3 = and i64 %2, 9223372036854775807
; CHECK-NEXT: %4 = bitcast i64 %3 to double
; CHECK-NEXT: %cmp = fcmp uge double %4, 0x3CB0000000000000
; CHECK-NEXT: %conv = zext i1 %cmp to i32
; CHECK-NOT: fcmp une
; CHECK-DAG: ret i32 %conv

  %cmp = fcmp une double %a, %b
  %conv = zext i1 %cmp to i32
  ret i32 %conv
}

define i32 @fp_cmp2(double %a, double %b) {
; CHECK-DAG: @fp_cmp2
; CHECK-NEXT: %fneg = fneg double %a
; CHECK-NEXT: %1 = fsub double %fneg, %b
; CHECK-NEXT: %2 = bitcast double %1 to i64
; CHECK-NEXT: %3 = and i64 %2, 9223372036854775807
; CHECK-NEXT: %4 = bitcast i64 %3 to double
; CHECK-NEXT: %cmp = fcmp olt double %4, 0x3CB0000000000000
; CHECK-NEXT: %conv = zext i1 %cmp to i32
; CHECK-NOT: fcmp oeq
; CHECK-DAG: ret i32 %conv

  %fneg = fneg double %a
  %cmp = fcmp oeq double %fneg, %b
  %conv = zext i1 %cmp to i32
  ret i32 %conv
}

define i32 @fp_cmp3(double %a, double %b) {
; CHECK-LABEL: @fp_cmp3
; CHECK-DAG %fneg = fneg double %a
; CHECK-NEXT %1 = fsub double %fneg, %b
; CHECK-NEXT %2 = bitcast double %1 to i64
; CHECK-NEXT %3 = and i64 %2, 9223372036854775807
; CHECK-NEXT %4 = bitcast i64 %3 to double
; CHECK-NEXT %cmp = fcmp uge double %4, 0x3CB0000000000000
; CHECK-NEXT %conv = zext i1 %cmp to i32
; CHECK-NOT: fcmp une
; CHECK-DAG ret i32 %conv

  %fneg = fneg double %a
  %cmp = fcmp une double %fneg, %b
  %conv = zext i1 %cmp to i32
  ret i32 %conv
}
```

Also, have you considered running the transformed output IR? I've left more specific suggestions inline in `test/ConvertFCmpEq.ll`.

File: lib/ConvertFCmpEq.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+extern "C" LLVM_ATTRIBUTE_WEAK ::llvm::PassPluginLibraryInfo
+llvmGetPassPluginInfo() {
+  return getConvertFCmpEqPluginInfo();
+}
+
+//-----------------------------------------------------------------------------
+// Legacy PM Registration
+//-----------------------------------------------------------------------------
+
+char FindFCmpEqWrapper::ID = 0;
```
[the remark this answers is not in the document]
Comment (jvstech) [reply lib/ConvertFCmpEq.cpp:209]: Egad that's embarrassing. Yes! You are correct.

File: include/ConvertFCmpEq.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#ifndef LLVM_TUTOR_CONVERT_FCMP_EQ_H
```
[the remark this answers is not in the document]
Comment (jvstech) [reply include/ConvertFCmpEq.h:1]: My apologies. I seriously thought I added comments to this, but on review, it was `ConvertFCmpEq.cpp`. I'll certainly add comments to this.

File: inputs/input_for_fcmp_eq.c [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  
+  return sqrt_impl(x, midpoint, hi);
+}
+
+double sqrt(double x)
+{
+  return sqrt_impl(x, 0, x / 2.0 + 1.0);
+}
+
+int compare_fp_values()
```
[the remark this answers is not in the document]
Comment (jvstech) [reply inputs/input_for_fcmp_eq.c:34]: I'm onboard!

File: lib/ConvertFCmpEq.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+// DESCRIPTION:
+//    Transformation pass which uses the results of the FindFCmpEq analysis pass
+//    to convert all equality-based floating point comparison instructions in a
+//    function to indirect, difference-based comparisons.
+//
+//    This example demonstrates how to couple an analysis pass with a 
+//    transformation pass, the use of statistics (the STATISTIC macro), and LLVM
+//    debugging operations (the LLVM_DEBUG macro and the llvm::dbgs() output 
+//    stream). It also demonstrates how instructions can be modified without
+//    without having to completely replace them.
```
[the remark this answers is not in the document]
Comment (jvstech) [reply lib/ConvertFCmpEq.cpp:14]: Ugh, thank you!

**Comment by banach-space** [comment]
Great to see this, thank you!

I think that this is ready to be merged, but preferably as one commit. Would you mind re-basing and squashing? This way you can decide what goes into the final commit message. Happy to do it for you if you are busy!

Thanks a ton for working on this!

**Comment by banach-space** [comment]
Merged manually: https://github.com/banach-space/llvm-tutor/commit/bbf7d97b3c38affa085d70aaa301b9a1252a8399

Closing. Thank you @jvstech !

**Pull request closed without merging**